sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala

```diff
@@ -1373,6 +1386,28 @@ class AdaptiveQueryExecSuite
     }
   }
 
+  test("SPARK-35585: Support propagate empty relation through project/filter") {
+    withSQLConf(SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "true",
+      SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "-1") {
+      val (plan1, adaptivePlan1) = runAdaptiveAndVerifyResult(
+        "SELECT key FROM testData WHERE key = 0 ORDER BY key, value")
+      assert(findTopLevelSort(plan1).size == 1)
+      assert(findTopLevelSort(adaptivePlan1).isEmpty)
+
+      val (plan2, adaptivePlan2) = runAdaptiveAndVerifyResult(
+        """
+          |SELECT t1.key, count(*) FROM testData t1
```

Comment:
We can use limit to test the filter case as Spark can't pushdown filter through limit
```
SELECT key FROM (SELECT key FROM testData WHERE key = 0 LIMIT 1)
```

Reply:
If `limit` is not the top-level node, we will introduce extra shuffle node for the `SinglePartition`. It will make plan complex.

Reply:
ok, we can use rand() then
```
SELECT key FROM (SELECT * FROM testData WHERE value = 'no_match' ORDER BY key) WHERE key > rand()
```

Reply:
ok done.